pyccl/base.py

```diff
@@ -298,3 +299,316 @@ def increment(self):
+            self.object_lock.lock()
+
+
+def unlock_instance(func=None, *, argv=0, mutate=True):
+    """Decorator that temporarily unlocks an instance of CCLObject.
+
+    Arguments:
+        func (``function``):
+            Function which changes one of its ``CCLObject`` arguments.
+        argv (``int``):
+            Which argument should be unlocked. Defaults to the first argument.
+            This should be a ``CCLObject``, or the decorator will do nothing.
+        mutate (``bool``):
+            If after the function ``instance_old != instance_new``, the
+            instance is mutated. If ``True``, the representation of the
+            object will be reset.
+    """
+    if func is None:
+        # called with parentheses
+        return functools.partial(unlock_instance, argv=argv, mutate=mutate)
+
+    @functools.wraps(func)
+    def wrapper(*args, **kwargs):
+        # Pick argument from list of `args` or `kwargs` as needed.
+        size = len(args)
+        arg = args[argv] if size > argv else list(kwargs.values())[argv-size]
+        with UnlockInstance(arg, mutate=mutate):
+            out = func(*args, **kwargs)
+        return out
+    return wrapper
+
+
+def Funlock(cls, name, mutate: bool):
+    """Helper to allow an instance to change or mutate when `name` is called.
+    """
+    func = vars(cls).get(name)
+    if func is not None:
+        newfunc = unlock_instance(mutate=mutate)(func)
+        setattr(cls, name, newfunc)
+
+
+class FancyRepr:
+    """Controls the usage of fancy ``__repr__` for ``CCLObjects."""
+    _enabled: bool = True
+    _classes: dict = {}
+
+    def __repr__(self):
+        return f"FancyRepr(enabled={self._enabled})"
+
+    def add(self, cls):
+        """Add class to the internal dictionary of fancy-repr classes."""
+        self._classes[cls] = cls.__repr__
+
+    def enable(self):
+        """Enable fancy representations if they exist."""
+        for cl, method in self._classes.items():
+            FancyRepr.bind_and_replace(cl, method)
+        self._enabled = True
+
+    def disable(self):
+        """Disable fancy representations and fall back to Python defaults."""
+        for cl in self._classes.keys():
+            cl.__repr__ = object.__repr__
+        self._enabled = False
+
+    @classmethod
+    def bind_and_replace(cls, cl, method):
+        """Bind a method to a class, and replace original with default.
+        This helper only works for binding and replacing ``__repr__`` methods
+        for ``CCLObjects``.
+        """
+        # If the class defines a custom `__repr__`, this will be the new
+        # `_repr` (which is cached). Decorator `cached_property` requires
+        # that `__set_name__` is called on it.
+        bmethod = functools.cached_property(method)
+        cl._repr = bmethod
+        bmethod.__set_name__(cl, "_repr")
+        # Fall back to using `__ccl_repr__` from `CCLObject`.
+        cl.__repr__ = cl.__ccl_repr__
+
+
+class _DisableGetMethod:
```

Comment:
See my comment about why this may not be a bad thing to have at the subclass level.